### PR TITLE
br: configure the httpclient for external storage

### LINKED ISSUE #46011: cannot reuse tcp connection for the external storage
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. run br truncate

### 2. What did you expect to see? (Required)
``` 
netstat | grep "TIME_WAIT"
0
```
### 3. What did you see instead (Required)
```
netstat | grep "TIME_WAIT"
11111
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #46011 

Problem Summary:
br use the default http client to connect the external storage, which doesn't have enough reuse-connection-pool for each host(the default value is 2). This leads to slow batch files processing and TCP connections exceeding the system configuration limit.

### What is changed and how it works?
set the `MaxIdleConnsPerHost` as the same as `MaxIdleConns`.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

before configuration
img_v2_6f0edf8a-a936-4ed6-aa18-77f5e663fa7g

after configuration
fd011f9c-050a-4c66-9a2a-57bfab0ab882

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/storage/storage.go [hunk 1]
```diff
@@ -198,3 +205,29 @@ func New(ctx context.Context, backend *backuppb.StorageBackend, opts *ExternalSt
...
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: defaultTransportDialContext(&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}),
+			ForceAttemptHTTP2:     true,
+			MaxIdleConns:          int(concurrency),
+			MaxIdleConnsPerHost:   int(concurrency),
+			IdleConnTimeout:       90 * time.Second,
```
Comment (BornChanger) [inline br/pkg/storage/storage.go:228]: Can we avoid to hard code them but copy attributes from http.DefaultTransport

Comment (Leavrth) [reply br/pkg/storage/storage.go:228]: the `http.DefaultTransport` is a interface type( interface `RoundTripper` ), so we only can force convert back to Transport and ignore the possible error.
And it has some fields, such as sync.Mutex and map, I'm worried that the copy leaves behind the risk of golang updating.

**Review by YuJuncen — APPROVED** [review]

File: br/pkg/storage/storage.go [hunk 2]
```diff
@@ -198,3 +205,29 @@ func New(ctx context.Context, backend *backuppb.StorageBackend, opts *ExternalSt
 		return nil, errors.Annotatef(berrors.ErrStorageInvalidConfig, "storage %T is not supported yet", backend)
 	}
 }
+
+// copy from `http.defaultTransportDialContext`
```
Comment (lance6716) [inline br/pkg/storage/storage.go:209]: another way is use go:linkname like https://github.com/pingcap/tidb/blob/5a305400a94464d226f95503c6a3068ef881e69d/br/pkg/lightning/backend/local/local_test.go#L752-L753

File: br/pkg/storage/storage.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/storage/storage.go:228]: I also like @BornChanger 's idea. We can add an unit test to detect the future implementation change of interface `RoundTripper` , and use `Transport.Clone` to correctly handle its private fields.

File: br/pkg/storage/storage_test.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	require.True(t, ok)
+	require.True(t, transport.MaxConnsPerHost == 0)
+	require.True(t, transport.MaxIdleConns > 0)
+}
+
+func TestDefaultHttpClient(t *testing.T) {
+	var concurrency uint = 128
+	transport, ok := storage.GetDefaultHttpClient(concurrency).Transport.(*http.Transport)
+	require.True(t, ok)
+	require.Equal(t, transport.MaxIdleConnsPerHost, int(concurrency))
```
Comment (lance6716) [inline br/pkg/storage/storage_test.go:24]: nit: except for `t`, the first argument is "expected" and the second one is "got", so we should swap them

**Review by lance6716 — COMMENTED** [review]
lgtm

**Pull request merged**